travis: add PR license check for missing/not-valid license files

0xc0170 opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

The goal: check license offenders in pull requests

This is similar to what astyle does in Travis. We get list of files being changed. Because scancode does not support list of files being scanned but rather a file or directory, we copy files to SCANCODE folder. Execute scancode license check in this folder and check for offenders.

The rules there are: code files must have a license and SDPX identifier. If they don't, we print these files with their licenses and request review (Travis fails).

There is currently no ignore list as we have for astyle. I believe all code files in our codebase should have valid license (even if they are 3rd party).

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

0xc0170 committed:
Add scancode evaluate script

The goal: check license offenders in pull request

This is similar to what astyle does in Travis. We get list of files being changed. Because scancode does not support list of files being scanned but rather a file or directory, we copy files to SCANCODE folder. Execute scancode license check in this folder and check for offenders.

The rules there are: code files must have a license and SDPX identifier. If they don't, we print these and ask for review.

It functions nicely there is just one workaround needed. SPDX is not always 100 percent correctly found, therefore we recheck file if no SPDX manually in the script. This proves to remove false positives.

SeppoTakalo reviewed (dismissed):
Looks OK, but is it missleading if our internal script that uses output of `scancode` is named as `scancode.py`?

Maybe `parse-scancode-output.py` or similar. 
But anyway, +1 for using a proper up to date tool for the purpose.

adbridge commented on tools/test/travis-ci/scancode-evaluate.py:
```diff
@@ -0,0 +1,105 @@
+
+    # find all licenses in the files, must be licensed and permissive
+    with open(file, 'r') as scancode_output:
+        results = json.load(scancode_output)
```
Could this fail to load the json if the json is invalid? ie should there by a try around the json.load?

0xc0170 replied:
Fixed, via ValueError catch

adbridge commented on tools/test/travis-ci/scancode-evaluate.py:
```diff
@@ -0,0 +1,105 @@
+                if file['type'] == 'directory':
+                    continue
+                if not file['licenses']:
+                    dict_file = FILE_OFFENDER_TEMPLATE
```
dict_file and dict_file['file'] are set the same in 3 different places would it not be better to set those two right at the start of the for loop ?

0xc0170 replied:
Fixed, one per loop

adbridge commented on tools/test/travis-ci/scancode-evaluate.py:
```diff
@@ -0,0 +1,105 @@
+    if args.file and os.path.isfile(args.file):
+            license_check(args.directory_name, args.file)
+    else:
+        sys.exit("Missing scancode json file to be checked")
```
Where is this going to run ? Should we be using logging rather than prints ?

0xc0170 replied:
I'll change them to logging.info

adbridge commented on tools/test/travis-ci/scancode-evaluate.py:
```diff
@@ -0,0 +1,109 @@
+def license_check(directory_name, file):
+
+    offenders = []
+    try:
```
Just thought of another possible exception - file doesn't exist.... So you may need nested trys there

0xc0170 replied:
That one is checked prior the function, let me add docs to the function

adbridge commented on tools/test/travis-ci/scancode-evaluate.py:
```diff
@@ -0,0 +1,109 @@
+    if offenders:
+        userlog.warning("Found files with missing license details, please review and fix")
+        for offender in offenders:
+            userlog.warning("File: " + offender['file']['path'][len(directory_name):] + " " + "reason: " + offender['reason'])
```
You can use C style string flags here ie "File: %s, %s", offender[][][], reason  ?

0xc0170 replied:
I would leave it as it is

adbridge commented on tools/test/travis-ci/scancode-evaluate.py:
```diff
@@ -0,0 +1,124 @@
+                    found_spdx = True
+
+            if not found_spdx:
+                # Issue reported reported here https://github.com/nexB/scancode-toolkit/issues/1913
```
reported reported

0xc0170 committed:
travis: exclude one file from scanning licenses

tools/test/toolchains/api_test.py is problematic as it contains code
for checking tools. This contains license code in strings - scancode things these are the licenses in the file.

We skip it checking in travis

0xc0170 committed:
scancode: skip non code files

For instance pdf, or binaries - catch them as an exception and skip for
additional SPDX check

0xc0170 committed:
travis: license check does not fail but report issues

Master branch contains lot of missing SPDX identifiers, we will clean them up but this will take some time. In the meantime, we should not increase the license missing files. Each PR will report if there is no license issue or positive number reported as Github status. Travis won't fail if there are issues. This will highlight the issues that anyone can fix.

As soon as master is clean, we can fix set_status and revert part of this commit.

0xc0170 committed:
scancode: add logger and parse it to get count

0xc0170 committed:
scancode: print files if any found

adbridge commented on drivers/CAN.h:
```diff
@@ -1,19 +1,4 @@
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
+
```
Where has the license header gone here ?

adbridge commented on drivers/internal/AsyncOp.h:
```diff
@@ -1,6 +1,5 @@
 /*
  * Copyright (c) 2018-2019, Arm Limited and affiliates.
- * SPDX-License-Identifier: Apache-2.0
  *
```
Why has SPDX line gone ?

0xc0170 replied:
testing purpose (to see what travis reportS) as stated earlier. I'l lrevert this with rebase
